tiled/client/cache.py

```diff
@@ -171,17 +187,23 @@ def __init__(
         self._filepath = filepath
         self._owner_thread = threading.current_thread().ident
         self._conn = _prepare_database(filepath, readonly)
+        self._lock = threading.Lock()
 
     def __repr__(self):
         return f"<{type(self).__name__} {str(self._filepath)!r}>"
 
     def write_safe(self):
-        """
-        Check that it is safe to write.
+        """Check that it is safe to write.
+
+        SQLite is not threadsafe for concurrent _writes_ unless the
+        underlying sqlite library was built with thread safety
+        enabled. Even still, it may be a good idea to use a thread
+        lock (``@with_thread_lock``) to prevent parallel writes.
 
-        SQLite is not threadsafe for concurrent _writes_.
         """
-        return threading.current_thread().ident == self._owner_thread
+        is_main_thread = threading.current_thread().ident == self._owner_thread
+        sqlite_is_safe = sqlite3.threadsafety > 2
+        return is_main_thread or sqlite_is_safe
```

Comment:
I'm not sure there is any guarantee about ordinality of the `threadsafety` value in `sqlite3`. It's probably safer and arguably clearer to use a named constant here.

```suggestion
        SERIALIZED = 3  # Could be defined in an enum elsewhere
        is_main_thread = threading.current_thread().ident == self._owner_thread
        sqlite_is_safe = sqlite3.threadsafety == SERIALIZED
        return is_main_thread or sqlite_is_safe
```

* https://docs.python.org/3/library/sqlite3.html#sqlite3.threadsafety
* https://github.com/python/cpython/blob/347acded845d07b70232cade8d576b6f0aaeb473/Modules/_sqlite/module.c#L521-L541